src/ProjectTemplates/BlazorTemplates.Tests/BlazorServerTemplateTest.cs

```diff
@@ -107,84 +124,98 @@ public async Task BlazorServerTemplateWorks_IndividualAuth(bool useLocalDB)
+                        await aspNetProcess.VisitInBrowserAsync(page);
+                        await TestBasicNavigation(page);
+                        await page.CloseAsync();
+                    }
                 }
                 else
                 {
-                    BrowserFixture.EnforceSupportedConfigurations();
+                    EnsureBrowserAvailable(browserKind);
                 }
             }
 
-
             using (var aspNetProcess = Project.StartPublishedProjectAsync())
             {
                 Assert.False(
                     aspNetProcess.Process.HasExited,
                     ErrorMessages.GetFailedProcessMessageOrEmpty("Run published project", Project, aspNetProcess.Process));
 
                 await aspNetProcess.AssertStatusCode("/", HttpStatusCode.OK, "text/html");
-                if (BrowserFixture.IsHostAutomationSupported())
+                if (Fixture.BrowserManager.IsAvailable(browserKind))
                 {
-                    aspNetProcess.VisitInBrowser(Browser);
-                    TestBasicNavigation();
+                    for (int i = 0; i < 100; i++)
+                    {
+                        var page = await browser.NewPageAsync();
+                        await aspNetProcess.VisitInBrowserAsync(page);
+                        await TestBasicNavigation(page);
+                        await page.CloseAsync(); 
+                    }
+                }
+                else
+                {
+                    EnsureBrowserAvailable(browserKind);
                 }
             }
         }
 
-        private void TestBasicNavigation()
+        private async Task TestBasicNavigation(IPage page)
         {
-            var retries = 3;
-            var connected = false;
-            do
-            {
-                try
-                {
-                    Browser.Contains("Information: WebSocket connected to",
-                        () => string.Join(Environment.NewLine, Browser.GetBrowserLogs(LogLevel.Info).Select(b => b.Message)));
-                    connected = true;
-                }
-                catch (TimeoutException) when(retries-- > 0)
-                {
-                    Browser.Navigate().Refresh();
-                }
-            } while (!connected && retries > 0);
+            var socket = (await page.WaitForEventAsync(PageEvent.WebSocket)).WebSocket;
+
+            // Start circuit
+            await socket.WaitForEventAsync(WebSocketEvent.FrameSent);
+
+            // Receive render batch
+            await socket.WaitForEventAsync(WebSocketEvent.FrameReceived);
+            await socket.WaitForEventAsync(WebSocketEvent.FrameSent);
+
+            // JS interop call to intercept navigation
+            await socket.WaitForEventAsync(WebSocketEvent.FrameReceived);
+            await socket.WaitForEventAsync(WebSocketEvent.FrameSent);
 
+            // Wait until SignalR sends a ping and we answer
```

Comment:
Why?

Reply:
lol, I'm seeing the socket getting disconnected and I suspect is because we click a link before blazor is "fully" ready. 

I'm not able to repro it locally (if you see, I'm running the automation 100 times in a loop) so this is my attempt to wait for "things to settle" before we start the test.

Reply:
For reference, some of the logs I see
```console
dbug: Microsoft.AspNetCore.Components.Server.Circuits.CircuitHost[206]
      The JS interop call with callback id '2' succeeded.
info: [Browser][https://localhost:52046/]
      [2021-02-10T21:20:50.897Z] Information: Normalizing '_blazor' to 'https://localhost:52046/_blazor'.
      (https://localhost:52046/_framework/blazor.server.js:0:5693)
error: [Browser]net::ERR_CONNECTION_REFUSED
error: [Browser][https://localhost:52046/]
      Failed to load resource: net::ERR_CONNECTION_REFUSED
      (https://localhost:52046/_blazor/negotiate?negotiateVersion=1:0:0)
warn: [Browser][https://localhost:52046/]
      [2021-02-10T21:20:53.219Z] Warning: Error from HTTP request. TypeError: Failed to fetch.
      (https://localhost:52046/_framework/blazor.server.js:0:5593)
error: [Browser][https://localhost:52046/]
      [2021-02-10T21:20:53.219Z] Error: Failed to complete negotiation with the server: TypeError: Failed to fetch
      (https://localhost:52046/_framework/blazor.server.js:0:5496)
error: [Browser][https://localhost:52046/]
      [2021-02-10T21:20:53.219Z] Error: Failed to start the connection: TypeError: Failed to fetch
      (https://localhost:52046/_framework/blazor.server.js:0:5496)
error: [Browser][https://localhost:52046/]
      [2021-02-10T21:20:53.220Z] Error: TypeError: Failed to fetch
      (https://localhost:52046/_framework/blazor.server.js:18:24519)
error: [Browser][https://localhost:52046/]
      [2021-02-10T21:20:53.220Z] Error: Error: Cannot send data if the connection is not in the 'Connected' State.
      (https://localhost:52046/_framework/blazor.server.js:18:24519)
info: [Browser][https://localhost:52046/]
      [2021-02-10T21:21:13.222Z] Information: Normalizing '_blazor' to 'https://localhost:52046/_blazor'.
      (https://localhost:52046/_framework/blazor.server.js:0:5693)
error: [Browser]net::ERR_CONNECTION_REFUSED
error: [Browser][https://localhost:52046/]
      Failed to load resource: net::ERR_CONNECTION_REFUSED
      (https://localhost:52046/_blazor/negotiate?negotiateVersion=1:0:0)
warn: [Browser][https://localhost:52046/]
      [2021-02-10T21:21:15.587Z] Warning: Error from HTTP request. TypeError: Failed to fetch.
      (https://localhost:52046/_framework/blazor.server.js:0:5593)
error: [Browser][https://localhost:52046/]
      [2021-02-10T21:21:15.587Z] Error: Failed to complete negotiation with the server: TypeError: Failed to fetch
      (https://localhost:52046/_framework/blazor.server.js:0:5496)
error: [Browser][https://localhost:52046/]
      [2021-02-10T21:21:15.587Z] Error: Failed to start the connection: TypeError: Failed to fetch
      (https://localhost:52046/_framework/blazor.server.js:0:5496)
error: [Browser][https://localhost:52046/]
      [2021-02-10T21:21:15.587Z] Error: TypeError: Failed to fetch
      (https://localhost:52046/_framework/blazor.server.js:18:24519)
error: [Browser][https://localhost:52046/]
      [2021-02-10T21:21:15.588Z] Error: Error: Cannot send data if the connection is not in the 'Connected' State.
      (https://localhost:52046/_framework/blazor.server.js:18:24519)
error: [Browser][https://localhost:52309/]
      [2021-02-10T21:21:20.356Z] Error: Connection disconnected with error 'Error: WebSocket closed with status code: 1006 ().'.
      (https://localhost:52309/_framework/blazor.server.js:0:5496)

```

Reply:
Looks like these logs are failing on negotiate which is before the websocket

Reply:
@BrennanConroy Yes, but I think that happens because we click on the counter button too fast and that causes a full page navigation to happen instead of a client-side navigation

Reply:
If you see the first line, the socket is already connected on the initial page

Reply:
I don't have access to the network trace on the PR but I've seen a similar issue when I was running locally and wasn't waiting long enoug

Reply:
Is it possible we need a more explicit signal that the Blazor Server client-side code is wired up and ready to go?

Reacting to the WebSockets traffic is an indicator that things have started happening, but I'm not sure it guarantees that `blazor.server.js` has got as far as activating the link click interception. There might still be some time gap where it can go wrong.

We are a bit limited in what we can do when testing the templates, as we can't add any test-specific logic to the application being tested. If we're keen to solve this, one option would be something like getting PlayWright to wait until the DOM contains some artifacts that are specific to interactive rendering, such as the comment marker tags we put around component outputs, or even explicitly looking for the delegating event handler registered on the root element.

If you think this is sufficiently unlikely to make a difference then I'm fine with leaving it until we see reliability issues. Might be good to clarify the comment in the test sources though to at least state what kind of test reliability issue this logic is meant to mitigate.

Reply:
@SteveSandersonMS at first I thought the reliability issues we were having were because we were not waiting enough, but after doing all this, what I'm seeing is that the connection to the server is lost in some cases.

I have not been able to reproduce this locally (I executed the whole E2E interaction 200 times in a loop several times with 100% success).

I think the amount of waiting that we do here is correct (and arguably we don't need to wait for the SignalR ping).

The reason for is:
* When the circuit starts (first frame sent) an initial render happens.
* That initial render renders the router which in turn triggers the interop wire up.
* The browser always receives the render batch before the interop wire up and always acks the first batch.
* Then the browser receives the interop call and acks the js interop invocation.
* At the point where we receive the second js interop invocation we are ready to go.

I think overall, waiting for frames is much more reliable than checking other secondary DOM or JS state that we can't clearly evaluate.

If we wanted to go that far, we could even crack open the payloads of the websockets to validate they were exactly what we were expecting.